RecoHGCal/TICL/plugins/PatternRecognitionbyKF.cc

```diff
@@ -0,0 +1,481 @@
+    if (countPos[i]) {
+      HGCDiskGeomDet* disk = new HGCDiskGeomDet(subdet, +1, layer, zsumPos[i]/countPos[i], rmin[i], rmax[i], radlen_[layer], xi_[layer]);
+      addDisk(disk);
+    }
+    if (countNeg[i]) {
+      HGCDiskGeomDet* disk = new HGCDiskGeomDet(subdet, -1, layer, zsumNeg[i]/countNeg[i], rmin[i], rmax[i], radlen_[layer], xi_[layer]);
+      addDisk(disk);  
+    }
+    layer++;
+  }
+}
+
+
+template <typename TILES>
+void PatternRecognitionbyKF<TILES>::fillHitMap(std::map<DetId,const HGCRecHit*>& hitMap,
+                                const HGCRecHitCollection& rechitsEE,
+                                const HGCRecHitCollection& rechitsFH,
+                                const HGCRecHitCollection& rechitsBH) const {
+  
+  hitMap.clear();
+  for (const auto& hit : rechitsEE) {
+    hitMap.emplace(hit.detid(), &hit);
+  }
+
+  for (const auto& hit : rechitsFH) {
+    hitMap.emplace(hit.detid(), &hit);
+  }
+
+  for (const auto& hit : rechitsBH) {
+    hitMap.emplace(hit.detid(), &hit);
+  }
+}
+
+template <typename TILES>
+std::vector<TrajectoryMeasurement> PatternRecognitionbyKF<TILES>::measurements(
+      const TrajectoryStateOnSurface &tsos, 
+      const MeasurementEstimator &mest, 
+      const TILES &tiles, 
+      int depth){
+  
+  std::vector<TrajectoryMeasurement> ret;
+
+  // define search window and get bins
+
+  float eta = tsos.globalPosition().eta();
+  float phi = tsos.globalPosition().phi();
+
+  float etaMin = eta - etaBinSize;
+  float etaMax = eta + etaBinSize;
+  float phiMin = phi - phiBinSize;
+  float phiMax = phi + phiBinSize;
+
+  auto bins = tiles[depth].searchBoxEtaPhi(etaMin, etaMax, phiMin, phiMax);
+
+  // loop over candidates
+
+  for (int ieta = bins[0]; ieta < bins[1]; ieta++) {
+    auto offset = ieta * nPhiBin;
+    for (int phi = bins[2]; phi < bins[3]; phi++) {
+      int iphi = ((phi % nPhiBin + nPhiBin) % nPhiBin);
+      if (!tiles[depth][offset + iphi].empty()) {
+        for(auto hit: tiles[depth][offset + iphi]) {
+          const auto rec = hitMap.find(hit)->second;
+          float energy = rec->energy();
+
+          GlobalPoint globalpoint = rhtools_.getPosition(hit);
+          LocalPoint localpoint = tsos.surface().toLocal(globalpoint);
+
+          auto hitptr = std::make_shared<HGCTrackingRecHit>(hit,localpoint,lerr[hit],energy);
+          auto mest_pair = mest.estimate(tsos, *hitptr);
+          if(mest_pair.first){
+            ret.emplace_back(tsos,hitptr,mest_pair.second);
+          }
+        }
+      }
+    }
+  }
+return ret;
+}
+
+template <typename TILES>
+void PatternRecognitionbyKF<TILES>::init(
+    const edm::Event& ev, const edm::EventSetup& es){
+
+    //Get Calo Geometry
+    if (es.get<CaloGeometryRecord>().cacheIdentifier() != geomCacheId_) {
+      geomCacheId_ = es.get<CaloGeometryRecord>().cacheIdentifier();
+      const CaloGeometry* geom = &es.getData(caloGeomToken_);
+      rhtools_.setGeometry(*geom);
+
+      makeDisks(8, 26, geom);
```

Comment:
as we have seen during our chat, total number of layers can be taken by the geom class!